### PR TITLE
delegate to the proper comparison operator

### LINKED ISSUE #62566: `c10::optional<T>` operators should delegate to corresponding operators on `T`
## 🚀 Feature
`c10::optional<T>` operators should delegate to corresponding operators on `T`. For example, `operator!=` should ultimately call `T::operator!=` rather than `T::operator==` as it currently does.

## Motivation

The lack of this makes our optional class possibly behave unexpectedly on a type `T` that does not have a full suite of comparison operators, e.g. if someone only implements `operator<` to work in `std::set`.

This also reduces distance from the `std::optional` implementation.

## Pitch

Extend tests to test comparison operations and then reimplement operators as needed. There are only a small number that are affected by this (e.g. `operator!=`, `operator>`, etc.).

## Alternatives

The current implementation works, but might behave somewhat unexpectedly in some subtle ways.

## Additional context

This is a small amount of work (< 1 day).

### PR DESCRIPTION
**Pull request by dagitses**
For example, call T::operator!= rather than !(x == y).

Fixes #62566

### CODE REVIEW

**Comment by mruberry** [comment]
@ezyang would you review or recommend a reviewer for this PR, too?

File: c10/util/Optional.h [hunk 1]
```diff
@@ -964,7 +964,7 @@ constexpr bool operator==(const optional<T>& x, const optional<T>& y) {
 
 template <class T>
 constexpr bool operator!=(const optional<T>& x, const optional<T>& y) {
-  return !(x == y);
+  return x ? *x != y : bool(y);
```
Comment (ezyang) [inline c10/util/Optional.h:967]: Why isn't the implementation here simply:

```
return bool(x) != bool(y) ? true : bool(x) == false ? false : *x != *y;
```

in symmetry with the implementation above?

Comment (dagitses) [reply c10/util/Optional.h:967]: The implementation above is not simpler and appears to be written by someone who doesn't really understand booleans.

It's almost always a code smell to use the equality operator on a boolean literal (e.g. `bool(x) == false`). And using a ternary operator to return boolean literals also makes little sense.

The nested ternary also makes the logic a bit hard to follow, IMO.

IMO we could simplify all the implementations, but I don't think that's a good use of time since we're fairly close to C++17.

This PR is mostly to address a subtle correctness issue where we diverge from the spec.

Comment (ezyang) [reply c10/util/Optional.h:967]: Writing standard library code in C++ is very subtle, so having good hygiene when making modifications like this is very important, in my opinion. If the modification you made is not equivalent to the old code, you will be in a horrible situation where `==` and `!=` have different semantics just because the code is written differently and you didn't update the old call sites. Better hygiene is if you opened a separate PR to switch all of the statements to the new style, and then made the bugfix here. (BTW, we're not *that* close to C++17; per https://github.com/pytorch/pytorch/issues/56055 we have to wait for CUDA 10 to die and it seems to be dying... slowly.)

Now I'm going to show you that the new implementation is not semantically equivalent to the old one. I extracted your implementation and the old implementation into this program:

```
#include <optional>

bool neq1(void* v, std::optional<void*> x) {
    return bool(x) ? v != *x : true;
}

extern "C" bool tgt(std::optional<void*> x, std::optional<void*> y) {
    return x ? neq1(*x, y) : bool(y);
}

extern "C" bool src(std::optional<void*> x, std::optional<void*> y) {
    return !(bool(x) != bool(y) ? false : bool(x) == false ? true : *x == *y);
}
```

Using clang 12.0.1 and flags `-O2 -g0 -std=c++17 -emit-llvm` (https://godbolt.org/z/Yqo7n7398) I extracted the LLVM IR and then compared src and tgt using Alive2, a translation validation tool (https://alive2.llvm.org/ce/z/g-T9Zz) to see if it is valid to replace the original (src) with your new implementation (tgt). It is not.

```
ERROR: Target is more poisonous than source

Example:
* %0 = poison
i8 %1 = #xff (255, -1)
* %2 = poison
i8 %3 = #x00 (0)

...

Source value: #x1 (1)
Target value: poison
```

Specifically, when the "has value" field of both input optionals is poisoned (i.e. undefined behavior), but  the values in the optionals are not, the original implementation is able to report that the optionals are definitely not equal. In your implementation, the result in this case is also undefined behavior (because you force a ternary on these values).

I also extracted GCC's implementation of optional (which doesn't involve ternaries and I do agree is much more pleasing than the old implementation) and showed that it is equivalent to the old code but not the new code.

```
extern "C" bool src(std::optional<void*> x, std::optional<void*> y) {
    return bool(x) != bool(y) || (bool(x) && *x != *y);
}
```

So this modification is just not right, and we shouldn't accept it.

**Review by ezyang — CHANGES_REQUESTED** [review]
the new implementation is not correct

**Comment by swolchok** [comment]
This should come with a now-passing test demonstrating that it fixes #62566 as well.

File: c10/util/Optional.h (hunk 1 above)
Comment (ezyang) [reply c10/util/Optional.h:967]: @swolchok mentions to me that it's pretty unlikely for has_value to become poisoned in practice, because the constructor should initialize it, and moving from standard types should leave it unspecified but valid.

Comment (dagitses) [reply c10/util/Optional.h:967]: The goal of this is not to have identical behavior to the previous implementation, though I did not intend to change it in this particular way. Very interesting!

I agree that it makes sense to make our implementations consistent, minimally from a style perspective. I don't believe that the priority of doing that is high though, **but it's easy enough and I'm willing to do it**. It would be less work than this discussion, but much less fun :).

However, unless I'm misunderstanding you, this implementation is not incorrect. It is well defined and correct for valid inputs. If the inputs are not valid, the output is also not valid. But that's ok, because the program is invalid since we already produced undefined behavior prior to entering this function.

IMO, it doesn't matter if an invalid input produces an invalid output. In fact I think it's counterproductive to consider invalid inputs when defining functions. Where does that line of thinking end?

I **do** think that there's merit in having the exact same implementation as a quality implementation of `std::optional` (libstdc++ or libc++) since we do eventually want to switch to those. **So I'm not opposed to replicating that (and that I'm willing to let trump anything).**

**Comment by dagitses** [comment]
> This should come with a now-passing test demonstrating that it fixes #62566 as well.

A negative-compilation example would be nice for this: do we have that infrastructure?

Otherwise a class that is instrumented to record operators called would do the trick.

Any other ideas?

**Comment by swolchok** [comment]
> A negative-compilation example would be nice for this: do we have that infrastructure?

I don't know, but I was thinking of just having a class that records which operators are called since that's the behavior you're trying to replicate.

File: c10/util/Optional.h (hunk 1 above)
Comment (ezyang) [reply c10/util/Optional.h:967]: > However, unless I'm misunderstanding you, this implementation is not incorrect. It is well defined and correct for valid inputs. If the inputs are not valid, the output is also not valid. But that's ok, because the program is invalid since we already produced undefined behavior prior to entering this function.

This is a misunderstanding of how UB works in LLVM. There are certainly classes of undefined behavior where, if it happens, all bets are off, the nuclear missiles can be fired. But there are also cases where there is undefined behavior, but not *that* undefined, and undef/poison are examples of this. undef you can think of as "reading from uninitialized memory", poison you can think of as "you overflowed an integer".  The Alive2 paper has a pretty succinct discussion of how these are handled inside LLVM https://www.cs.utah.edu/~regehr/alive2-pldi21.pdf

Comment (ezyang) [reply c10/util/Optional.h:967]: Undefined behavior is a runtime concept: the text of a program doesn't have undefined behavior, only the execution of a program with some inputs is undefined behavior. (This is why the program `x+1` isn't UB, even though x could be MAX_INT and overflow). But the life of a compiler is all about moving instructions around and generally changing the source program, and so a compiler might want to, for example, hoist `x+1` out of a loop and "Yes, OK, maybe that could have caused UB, but please don't blow up immediately because in the source program this only would have happened in the loop, so please blow up *later* if you actually would have ended up in the loop", which is where the concepts of poison and undef come from.

To put it differently, reasoning about the behavior of a function under undef/poison helps us reason about how the program interacts with other code that might have UB (but then again, might not! But the compiler doesn't know what input it's actually going to run on, it's got to optimize compatibly with all possible inputs.)

**Comment by dagitses** [comment]
> the new implementation is not correct

per-offline discussion, we decided to replace the implementations with what libstdc++ does. I have not yet done that.

However, this morning I spent 30m hacking together a replacement of this (and `c10::variant`) with the corresponding `absl` replacements. It mostly works except for our optimization of `c10::optional<ArrayRef<T>>` introduced by @swolchok in #59333. Scott: was this a meaningful performance win such that we'd want to keep it by other means?

As far as I can tell, that's the only blocker that would prevent us from ripping this out when we can build in C++17 mode.

@ezyang what do you think about replacing these implementations with Abseil, Boost, or Folly (does not include variant, as far as i can tell) in the short-term? My take is that if our goal is to replace them with the standard implementations as soon as we can, taking an intermediate step that closely adheres to the standard will prevent future issues from creeping in. And we only pay a small amount for that defense: whatever it takes to integrate the third party dependency that we use.

**Comment by swolchok** [comment]
> Scott: was this a meaningful performance win such that we'd want to keep it by other means?

The original diff came with an internal measurement that suggests we want to keep it, yes. Could this be as simple as

```
namespace c10 {
template <typename T>
class Optional : public absl::optional<T> {};

template <typename T>
class Optional<c10::ArrayRef<T>> : public c10::OldOptional<T> {};
} // namespace c10
```

where `OldOptional` is the current implementation?

**Comment by swolchok** [comment]
Equivalently, we could just specialize `absl::optional<c10::ArrayRef<T>>`, and then change the specialization to be a `std::optional` specialization on that glorious day when we can use C++17.

I am surprised that we can just add abseil to PyTorch, though. I thought we didn't want to add additional third-party dependencies.

**Comment by dagitses** [comment]
> > Scott: was this a meaningful performance win such that we'd want to keep it by other means?
> 
> The original diff came with an internal measurement that suggests we want to keep it, yes. Could this be as simple as

I can't find the measurement in the diff. Can you help me?

> 
> ```
> namespace c10 {
> template <typename T>
> class Optional : public absl::optional<T> {};
> 
> template <typename T>
> class Optional<c10::ArrayRef<T>> : public c10::OldOptional<T> {};
> } // namespace c10
> ```
> 
> where `OldOptional` is the current implementation?

Maybe. But that would require keeping our implementation around, which wouldn't be ideal. IMO, I'd rather see a class like `OptionalArrayRef` that embeds this logic if it's not part of the standard implementation and there's not a clean way to add it via specialization. I don't have a horse in this race though.

**Comment by dagitses** [comment]
> Equivalently, we could just specialize `absl::optional<c10::ArrayRef<T>>`, and then change the specialization to be a `std::optional` specialization on that glorious day when we can use C++17.
> 
> I am surprised that we can just add abseil to PyTorch, though. I thought we didn't want to add additional third-party dependencies.

I think it's a good idea to have a very high bar for adding any third party dependencies. But it should always be weighed against the alternative, e.g. maintaining a subpar equivalent. Or arguably in our case, a superior equivalent that gives us enhanced performance.

**Comment by swolchok** [comment]
> I can't find the measurement in the diff. Can you help me?

It's in the test plan on Phabricator.
